### ISSUE TITLE
Bug: The quickfix extension ignores the theme and uses hard-coded highlight names

### ISSUE THREAD

**Issue by kmoschcau** [nvim-lualine/lualine.nvim#1029]
### Self Checks

- [X] I'm using the latest lualine.
- [X] I didn't find the issue in existing issues or PRs.

### How to reproduce the problem

Have a custom theme defined that sets colors for the `a` section, then open a quickfix or location list.

### Expected behaviour

The color for the first section is taken from the theme.

### Actual behaviour

The color is set to either the `Identifier` or `Constant`. Worse, it takes the foreground color of these highlight groups and uses them as a background color. Additionally it uses a deprecated API method to retrieve those highlights. Which also seems to have the effect that I get cterm values as colors, that are not even defined as such in my nvim theme.

This shows the difference between the theme's colors (in the status line of a buffer) and the colors of the quickfix window.
![difference](https://user-images.githubusercontent.com/9497302/235090585-dacae7d6-b3f4-4cea-8a1f-f7b84d330f0e.png)

This shows the theme highlight for `Identifier`.
![Identifier](https://user-images.githubusercontent.com/9497302/235091094-dfbd9ac0-2e64-4a27-b7ea-b99d05f7c5c0.png)

This shows the theme highlight for `Constant`.
![Constant](https://user-images.githubusercontent.com/9497302/235091368-16b081fa-d064-4f33-a886-a1afd6c016f4.png)

### Minimal config to reproduce the issue

N/A, just look at the source code of the extension.

### Additional information

Using `vim.api.nvim_get_hl_by_name('Constant', false).foreground` in the extension seems to fetch a cterm color, but again it's something completely different than expected. The theme I am using defines both GUI colors and fitting cterm codes and it certainly isn't what's returned by that function.
I am willing to look into how to improve this myself, but I need to do some research first.

**Comment by zankich** [comment]
I ran into the same issue and can see the incorrect quickfix highlights when listing the highlight groups
```
lualine_a_17_normal xxx guifg=#e2e2e2 guibg=#00ffff
lualine_a_17_insert xxx guifg=#e2e2e2 guibg=#00ffff
lualine_a_17_visual xxx guifg=#e2e2e2 guibg=#00ffff
lualine_a_17_replace xxx guifg=#e2e2e2 guibg=#00ffff
lualine_a_17_command xxx guifg=#e2e2e2 guibg=#00ffff
lualine_a_17_terminal xxx guifg=#e2e2e2 guibg=#00ffff
lualine_a_17_inactive xxx guifg=#777777 guibg=#00ffff
```

For some reason the `17` will change when you restart nvim so I wrote a brute force function which will look for any hl group that matches that pattern and link the quickfix highlight group to the other `lualine_a_` highlight values.

You can add something like this to your config until this bug is fixed.
```lua
lualine_a = {
  {
    "mode",
    color = function()
      for _, group in ipairs(vim.fn.getcompletion("lualine_a_*_*$", "highlight")) do
        if string.match(group, "^lualine_a_%d+_.+") then
          vim.api.nvim_set_hl(0, group, { link = string.gsub(group, "^lualine_a_%d+", "lualine_a") })
        end
      end
    end,
  },
}
```

**Comment by kmoschcau** [comment]
@gennaro-tedesco from my understanding and looking at how other extensions did it, it should have been enough to just not set the color property on the section at all. Lualine would already pick the color from the theme on its own.

**Comment by gennaro-tedesco** [comment]
> @gennaro-tedesco from my understanding and looking at how other extensions did it, it should have been enough to just not set the color property on the section at all. Lualine would already pick the color from the theme on its own.

...but then you would have both quickfix and location list sharing the same colour: the point of the extension is to have the section highlighted differently according to whether you open the quicklist or the location list.

**Comment by kmoschcau** [comment]
In that case the now included behavior seems like an odd design choice. Especially as it is the default behavior now, making it hard for users to change the highlight. This seems like the PR should have stayed as a work in progress and the placeholder highlights replaced with a solution before being merged.

In any case, I would not just bind this to some built-in highlight that has nothing to do with lualine by default. Maybe check for the existence of a specific highlight group first, then fall back to something hard-coded like the normal "a" section highlight. This would allow people to have a different highlight, but not break it for others.

Looking at the lualine wiki entry for writing themes, it seems like the best way to make this interactible for themes writers would be to introduce a new mode or similar.

**Comment by gennaro-tedesco** [comment]
> In that case the now included behavior seems like an odd design choice. 

Could you please explain which choice seems odd? The purpose of the extension is _exactly_ to create a function that assigns two different colour groups to quickfix and location list: now how these two colours are picked is a technical detail. As explained in the PR text these should be dynamically pass by the user: since I do not know whether it is possible at all to achieve such behaviour I left it as "todo"/comment for the repository owner or whomever knows the solution to it.

> making it hard for users to change the highlight.

?? You can literally just change [this line](https://github.com/nvim-lualine/lualine.nvim/blob/05d78e9fd0cdfb4545974a5aa14b1be95a86e9c9/lua/lualine/extensions/quickfix.lua#L20-L21) to your colours of choice.

> This seems like the PR should have stayed as a work in progress and the placeholder highlights replaced with a solution before being merged.

If the PR had stayed as work in progress, you could still not use it _unless_ you would fork the repository and change the colours yourself, which you can still do in fact now. 

Again, I understand your sentiment, however the solution to this is as simple as opening a new PR with the correct method to pass colours dynamically. Until then (or assuming such behaviour cannot be achieved altogether) changing the colours ad-hoc to adapt to one's colour scheme does not seem the end of the world to me :)

**Comment by kmoschcau** [comment]
What I mean by it seeming odd, is that this extension did already exist before and its sole job was to provide a better formatted status line for quickfix and location list windows. It now also changing highlights to something that can't easily be changed by the end user is an odd design choice.

And what I mean by it can't be easily changed by the end user or it is hard for users to change the highlights is, that it can't currently be done via the already existing configuration API of this plugin (lualine as a whole, not just this lualine extension). You are now expecting users to change the source files of a plugin, which they likely already installed via a plugin manager, just to change some highlights. This should be configurable via the API in the user's init files and shouldn't require them to edit plugin source code.

And yes it is my opinion that this should have stayed unavailable to users until it was configurable and working as intended before being merged and released to "production".

As for fixing this, I already laid out some suggestions in my previous post. That's all I can think of without digging into lualine's code any further.

**Comment by przepompownia** [comment]
Currently both values 
https://github.com/nvim-lualine/lualine.nvim/blob/45e27ca739c7be6c49e5496d14fcf45a303c3a63/lua/lualine/extensions/quickfix.lua#L20-L21
are `nil` for me.

After chaning deprecated (since nvim 0.9) form to 
```lua
local qf_colours = {
  ll = vim.api.nvim_get_hl(0, { name = 'Constant' }).fg,
  qf = vim.api.nvim_get_hl(0, { name = 'Identifier' }).fg,
}
```
both values can be > 255 so they do not pass
https://github.com/nvim-lualine/lualine.nvim/blob/45e27ca739c7be6c49e5496d14fcf45a303c3a63/lua/lualine/highlight.lua#L62-L77

Moreover, it turns out that https://github.com/nvim-lualine/lualine.nvim/blob/45e27ca739c7be6c49e5496d14fcf45a303c3a63/lua/lualine/highlight.lua#L108

still using concatenated command `hi!` passed to vim.cmd, not `nvim_set_hl`.

**Comment by przepompownia** [comment]
Do we really need compatibility with nvim 0.5?

**Comment by coffebar** [comment]
```lua
-- this line in extension now is not working
print(vim.api.nvim_get_hl_by_name("Constant", false).foreground) -- 13
-- this is working
print(vim.api.nvim_get_hl(0, { name = "Constant" }).fg) -- 12303...
```
What about PR? 

Should not be merged #1101 ?

**Comment by shadmansaleh** [comment]
I think changing the color based if it's location list or quickfix in lualine was a bad idea. it's simple enough to have change color if anyone wants to in userconfig side. picking colors that'd work with all themes is just annoying. on top of that hardcoding it makes no sense. I'll revert that commit

anyone still wanting the current behavior can do something like the following. even set there own colors in `qf_colors`

```lua
local quickfix_custom = require('lualine.extensions.quickfix');
quickfix_custom.sections.lualine_a[1].color = function()
  local is_loclist = vim.fn.getloclist(0, { filewinid = 1 }).filewinid ~= 0
  local qf_colours = {
    ll = vim.api.nvim_get_hl_by_name('Constant', false).foreground,
    qf = vim.api.nvim_get_hl_by_name('Identifier', false).foreground,
  }
  return  { bg = is_loclist and qf_colours['ll'] or qf_colours['qf']}
end

require('lualine').setup {
     extensions = { quickfix_custom, 'fugitive', 'nvim-dap-ui'}
}
```

**Comment by shadmansaleh** [comment]
closed with #1118
